File: typescript/base/src/main/vsdk/toolkit/io/wrapper/FileUncompressWrapper.ts
import { File } from "../../../../java/io/File";
import { FileInputStream } from "../../../../java/io/FileInputStream";
import { FileOutputStream } from "../../../../java/io/FileOutputStream";
import { InputStream } from "../../../../java/io/InputStream";
import { OutputStream } from "../../../../java/io/OutputStream";
import { Logger } from "../../common/logging/Logger";
import { PipeInputStream } from "./PipeInputStream";
import { PipeOutputStream } from "./PipeOutputStream";
import { StreamOpenMode } from "./StreamOpenMode";

type CommandHolder = {
  value: string;
};

export class FileUncompressWrapper {
  public static openInputStreamCompressWrapper(fileName: string | null, isPipe: number[] | null): InputStream | null {
    if (isPipe !== null && isPipe.length > 0) {
      isPipe[0] = 0;
    }
    if (FileUncompressWrapper.isInvalidFileName(fileName)) {
      return null;
    }
    const safeFileName = fileName as string;

    const commandLength = safeFileName.length + 20;
    const command: CommandHolder = { value: "" };
    const pipeFlag = FileUncompressWrapper.buildPipeCommand(safeFileName, StreamOpenMode.READ, command, commandLength);

    let stream: InputStream | null = null;
    if (pipeFlag) {
      stream = FileUncompressWrapper.openPipeInputStream(command.value);
    }
    else {
      const file = new File(safeFileName);
      const canOpen = file.exists() && file.canRead() && file.isFile();
      file.dispose();
      if (canOpen) {
        stream = new FileInputStream(safeFileName);
      }
    }

    if (stream === null) {
      Logger.error(null, "Can't open file '%s' for %s", safeFileName, FileUncompressWrapper.modeToLogAction(StreamOpenMode.READ));
      if (isPipe !== null && isPipe.length > 0) {
        isPipe[0] = 0;
      }
      return null;
    }

    if (isPipe !== null && isPipe.length > 0) {
      isPipe[0] = pipeFlag ? 1 : 0;
    }
    return stream;
  }

  public static openOutputStreamCompressWrapper(fileName: string | null, isPipe: number[] | null): OutputStream | null {
    if (isPipe !== null && isPipe.length > 0) {
      isPipe[0] = 0;
    }
    if (FileUncompressWrapper.isInvalidFileName(fileName)) {
      return null;
    }
    const safeFileName = fileName as string;

    const commandLength = safeFileName.length + 20;
    const command: CommandHolder = { value: "" };
    const pipeFlag = FileUncompressWrapper.buildPipeCommand(safeFileName, StreamOpenMode.WRITE, command, commandLength);

    let stream: OutputStream | null = null;
    if (pipeFlag) {
      stream = FileUncompressWrapper.openPipeOutputStream(command.value);
    }
    else {
      const file = new File(safeFileName);
      const isDirectory = file.isDirectory();
      file.dispose();
      if (!isDirectory) {
        stream = new FileOutputStream(safeFileName);
      }
    }

    if (stream === null) {
      Logger.error(null, "Can't open file '%s' for %s", safeFileName, FileUncompressWrapper.modeToLogAction(StreamOpenMode.WRITE));
      if (isPipe !== null && isPipe.length > 0) {
        isPipe[0] = 0;
      }
      return null;
    }

    if (isPipe !== null && isPipe.length > 0) {
      isPipe[0] = pipeFlag ? 1 : 0;
    }
    return stream;
  }

  public static closeInputStream(stream: InputStream | null): void {
    if (stream === null) {
      return;
    }
    try {
      stream.close();
    }
    catch (_error) {
    }
  }

  public static closeOutputStream(stream: OutputStream | null): void {
    if (stream === null) {
      return;
    }
    try {
      stream.close();
    }
    catch (_error) {
    }
  }

  private static modeToLogAction(mode: StreamOpenMode): string {
    return mode === StreamOpenMode.READ ? "reading" : "writing";
  }

  private static isInvalidFileName(fileName: string | null): boolean {
    if (fileName === null || fileName.length === 0 || fileName.endsWith("/")) {
      return true;
    }
    return false;
  }

  private static buildPipeCommand(
    fileName: string | null,
    openMode: StreamOpenMode,
    command: CommandHolder | null,
    commandLength: number
  ): boolean {
    if (fileName === null || command === null || commandLength <= 0) {
      return false;
    }

    command.value = "";

    if (fileName.charAt(0) === "|") {
      command.value = fileName.substring(1);
      return true;
    }

    const dot = fileName.lastIndexOf(".");
    const ext = dot >= 0 ? fileName.substring(dot) : null;

    if (ext === ".gz") {
      if (openMode === StreamOpenMode.READ) {
        command.value = `gunzip < ${fileName}`;
      }
      else {
        command.value = `gzip > ${fileName}`;
      }
    }
    else if (ext === ".Z") {
      if (openMode === StreamOpenMode.READ) {
        command.value = `uncompress < ${fileName}`;
      }
      else {
        command.value = `compress > ${fileName}`;
      }
    }
    else if (ext === ".bz") {
      if (openMode === StreamOpenMode.READ) {
        command.value = `bunzip < ${fileName}`;
      }
      else {
        command.value = `bzip > ${fileName}`;
      }
    }
    else if (ext === ".bz2") {
      if (openMode === StreamOpenMode.READ) {
        command.value = `bunzip2 < ${fileName}`;
      }
      else {
        command.value = `bzip2 > ${fileName}`;
      }
    }
    else {
      return false;
    }
    return true;
  }

  private static openPipeInputStream(command: string): InputStream | null {
    const pipeStream = new PipeInputStream(command);
    if (!pipeStream.isOpen()) {
      return null;
    }
    return pipeStream;
  }

  private static openPipeOutputStream(command: string): OutputStream | null {
    const pipeStream = new PipeOutputStream(command);
    if (!pipeStream.isOpen()) {
      return null;
    }
    return pipeStream;
  }
}

File: cpp/base/src/main/vsdk/toolkit/io/wrapper/FileUncompressWrapper.cpp
#include <cstring>

#include "java/io/FileInputStream.h"
#include "java/io/FileOutputStream.h"
#include "java/util/Formatter.h"
#include "vsdk/toolkit/common/logging/Logger.h"
#include "vsdk/toolkit/io/wrapper/FileUncompressWrapper.h"
#include "vsdk/toolkit/io/wrapper/PipeInputStream.h"
#include "vsdk/toolkit/io/wrapper/PipeOutputStream.h"
#include "vsdk/toolkit/io/wrapper/StreamOpenMode.h"

const char *
FileUncompressWrapper::modeToLogAction(StreamOpenMode mode) {
    return mode == StreamOpenMode::READ ? "reading" : "writing";
}

bool
FileUncompressWrapper::isInvalidFileName(const char *fileName) {
    if ( fileName == nullptr || fileName[0] == '\0' || fileName[strlen(fileName) - 1] == '/' ) {
        return true;
    }
    return false;
}

bool
FileUncompressWrapper::buildPipeCommand(const char *fileName, StreamOpenMode openMode, char *command, int commandLength) {
    if ( fileName == nullptr || command == nullptr || commandLength <= 0 ) {
        return false;
    }

    if ( fileName[0] == '|' ) {
        java::Formatter::format(command, commandLength, "%s", &fileName[1]);
        return true;
    }

    const char *ext = strrchr(fileName, '.');
    if ( ext && std::strcmp(ext, ".gz") == 0 ) {
        if ( openMode == StreamOpenMode::READ ) {
            java::Formatter::format(command, commandLength, "gunzip < %s", fileName);
        } else {
            java::Formatter::format(command, commandLength, "gzip > %s", fileName);
        }
    } else if ( ext && std::strcmp(ext, ".Z") == 0 ) {
        if ( openMode == StreamOpenMode::READ ) {
            java::Formatter::format(command, commandLength, "uncompress < %s", fileName);
        } else {
            java::Formatter::format(command, commandLength, "compress > %s", fileName);
        }
    } else if ( ext && std::strcmp(ext, ".bz") == 0 ) {
        if ( openMode == StreamOpenMode::READ ) {
            java::Formatter::format(command, commandLength, "bunzip < %s", fileName);
        } else {
            java::Formatter::format(command, commandLength, "bzip > %s", fileName);
        }
    } else if ( ext && std::strcmp(ext, ".bz2") == 0 ) {
        if ( openMode == StreamOpenMode::READ ) {
            java::Formatter::format(command, commandLength, "bunzip2 < %s", fileName);
        } else {
            java::Formatter::format(command, commandLength, "bzip2 > %s", fileName);
        }
    } else {
        return false;
    }
    return true;
}

java::InputStream *
FileUncompressWrapper::openPipeInputStream(const char *command) {
    PipeInputStream *pipeStream = new PipeInputStream(command);
    if ( !pipeStream->isOpen() ) {
        delete pipeStream;
        return nullptr;
    }
    return pipeStream;
}

java::OutputStream *
FileUncompressWrapper::openPipeOutputStream(const char *command) {
    PipeOutputStream *pipeStream = new PipeOutputStream(command);
    if ( !pipeStream->isOpen() ) {
        delete pipeStream;
        return nullptr;
    }
    return pipeStream;
}

java::InputStream *
FileUncompressWrapper::openInputStreamCompressWrapper(const char *fileName, int *isPipe) {
    if ( isPipe != nullptr ) {
        *isPipe = 0;
    }
    if ( isInvalidFileName(fileName) ) {
        return nullptr;
    }

    const int commandLength = static_cast<int>(strlen(fileName)) + 20;
    char *command = new char[commandLength];
    const bool pipeFlag = buildPipeCommand(fileName, StreamOpenMode::READ, command, commandLength);

    java::InputStream *stream = nullptr;
    if ( pipeFlag ) {
        stream = openPipeInputStream(command);
    } else {
        java::File file(fileName);
        if ( file.exists() && file.canRead() && file.isFile() ) {
            stream = new java::FileInputStream(fileName);
        }
    }
    delete[] command;

    if ( stream == nullptr ) {
        Logger::error(nullptr, "Can't open file '%s' for %s", fileName, modeToLogAction(StreamOpenMode::READ));
        if ( isPipe != nullptr ) {
            *isPipe = 0;
        }
        return nullptr;
    }

    if ( isPipe != nullptr ) {
        *isPipe = pipeFlag ? 1 : 0;
    }
    return stream;
}

java::OutputStream *
FileUncompressWrapper::openOutputStreamCompressWrapper(const char *fileName, int *isPipe) {
    if ( isPipe != nullptr ) {
        *isPipe = 0;
    }
    if ( isInvalidFileName(fileName) ) {
        return nullptr;
    }

    const int commandLength = static_cast<int>(strlen(fileName)) + 20;
    char *command = new char[commandLength];
    const bool pipeFlag = buildPipeCommand(fileName, StreamOpenMode::WRITE, command, commandLength);

    java::OutputStream *stream = nullptr;
    if ( pipeFlag ) {
        stream = openPipeOutputStream(command);
    } else {
        java::File file(fileName);
        if ( file.canWrite() && !file.isDirectory() ) {
            stream = new java::FileOutputStream(fileName);
        }
    }
    delete[] command;

    if ( stream == nullptr ) {
        Logger::error(nullptr, "Can't open file '%s' for %s", fileName, modeToLogAction(StreamOpenMode::WRITE));
        if ( isPipe != nullptr ) {
            *isPipe = 0;
        }
        return nullptr;
    }

    if ( isPipe != nullptr ) {
        *isPipe = pipeFlag ? 1 : 0;
    }
    return stream;
}

void
FileUncompressWrapper::closeInputStream(java::InputStream *stream) {
    if ( stream == nullptr ) {
        return;
    }
    stream->close();
    delete stream;
}

void
FileUncompressWrapper::closeOutputStream(java::OutputStream *stream) {
    if ( stream == nullptr ) {
        return;
    }
    stream->close();
    delete stream;
}
